client.go

```diff
@@ -41,38 +41,55 @@ func init() {
 	grpclog.SetLogger(log.New(ioutil.Discard, "", log.LstdFlags))
 }
 
-type NewClientOpts func(c *Client) error
+type clientOpts struct {
```

Comment:
Do we need to export it though? Everything we add here should have a corresponding `With` function, having this unexported prevents callers from implementing the these option functions themselves rather than using the ones already existing ones. Correspondingly everything in this struct must then be exported to be useful.  Most uses of `With*` patterns I have seen use unexported types here to have better control of the configuration (such as grpc, docker distribution)